File: packages/sdk-py/rein/x402/encoder.py
"""Build the value of the ``X-Payment`` HTTP header per the x402 spec."""

from __future__ import annotations

import base64
import json
from typing import Any, Optional

from rein.x402.parser import Requirement


def _b64u_encode(s: str) -> str:
    """Encode a UTF-8 string as base64url without padding."""
    return base64.urlsafe_b64encode(s.encode("utf-8")).decode("ascii").rstrip("=")


def encode_payment_header(
    *,
    requirement: Requirement,
    signature: str,
    transaction_base58: Optional[str] = None,
    receipt_pda: Optional[str] = None,
    extra: Optional[dict[str, Any]] = None,
) -> str:
    """Per Coinbase x402 v0.3 envelope:
    ``base64url(JSON({ scheme, network, payload: { signature, transaction?, receiptPda?, ... } }))``.
    """
    payload: dict[str, Any] = {"signature": signature}
    if transaction_base58 is not None:
        payload["transaction"] = transaction_base58
    if receipt_pda is not None:
        payload["receiptPda"] = receipt_pda
    if extra:
        payload.update(extra)

    envelope = {
        "scheme": requirement.scheme,
        "network": requirement.network,
        "payload": payload,
    }
    return _b64u_encode(json.dumps(envelope, separators=(",", ":")))

File: packages/sdk-ts/src/x402/encoder.ts
import type { Requirement } from './parser';

/**
 * Build the value of the `X-Payment` HTTP header per the x402 spec.
 *
 * Format (Coinbase v0.3, the v1 default):
 *   base64url(JSON({
 *     scheme,                  // mirrored from the requirement
 *     network,                 // mirrored from the requirement
 *     payload: {
 *       signature: '<base58 tx signature>',
 *       transaction: '<base58 serialized tx>'?,   // optional for some schemes
 *       receiptPda: '<base58 PDA>'?,              // REIN-specific
 *     }
 *   }))
 *
 * Facilitators may extend the payload — we pass through any extras the caller
 * supplies via `extra`.
 */
export function encodePaymentHeader(args: {
  requirement: Requirement;
  signature: string;
  transactionBase58?: string;
  receiptPda?: string;
  extra?: Record<string, unknown>;
}): string {
  const payload: Record<string, unknown> = {
    signature: args.signature,
  };
  if (args.transactionBase58) payload['transaction'] = args.transactionBase58;
  if (args.receiptPda) payload['receiptPda'] = args.receiptPda;
  if (args.extra) Object.assign(payload, args.extra);

  const envelope = {
    scheme: args.requirement.scheme,
    network: args.requirement.network,
    payload,
  };
  const json = JSON.stringify(envelope);
  return b64uEncode(json);
}

function b64uEncode(s: string): string {
  if (typeof globalThis.btoa === 'function') {
    const bytes = new TextEncoder().encode(s);
    let bin = '';
    for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
    return globalThis
      .btoa(bin)
      .replace(/=+$/g, '')
      .replace(/\+/g, '-')
      .replace(/\//g, '_');
  }
  return Buffer.from(s, 'utf-8')
    .toString('base64')
    .replace(/=+$/g, '')
    .replace(/\+/g, '-')
    .replace(/\//g, '_');
}
